client/post-editor/media-modal/detail/detail-preview-image.jsx

```diff
@@ -1,36 +1,104 @@
 /**
  * External dependencies
  */
-import React, { PropTypes } from 'react';
+import React, { PropTypes, Component } from 'react';
+import { noop } from 'lodash';
+import classNames from 'classnames';
 
 /**
  * Internal dependencies
  */
 import ImagePreloader from 'components/image-preloader';
 import Spinner from 'components/spinner';
-import MediaUtils from 'lib/media/utils';
+import { url, isItemBeingUploaded } from 'lib/media/utils';
 
-export default React.createClass( {
-	displayName: 'EditorMediaModalDetailPreviewImage',
-
-	propTypes: {
+export default class EditorMediaModalDetailPreviewImage extends Component {
+	static propTypes = {
 		site: PropTypes.object,
-		item: PropTypes.object.isRequired
-	},
+		item: PropTypes.object.isRequired,
+	};
+
+	static defaultProps = {
+		onLoad: noop
+	};
+
+	constructor( props ) {
+		super( props );
+
+		this.onImagePreloaderLoad = this.onImagePreloaderLoad.bind( this );
+		this.state = { loading: false };
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.item.URL === nextProps.item.URL ) {
+			return null;
+		}
+
+		this.setState( { loading: true } );
+	}
+
+	onImagePreloaderLoad() {
+		this.setState( { loading: false } );
+		this.props.onLoad();
+	}
 
 	render() {
-		const src = MediaUtils.url( this.props.item, {
+		const src = url( this.props.item, {
 			photon: this.props.site && ! this.props.site.is_private
 		} );
+		const uploading = isItemBeingUploaded( this.props.item );
+		const loading = this.state.loading;
+		const isBlob = /^blob/.test( src );
+
+		// Let's add special classes to differentiate
+		// the different states that an image could have.
+		//
+		// - `is-uploading` when the image is being uploaded
+		//    from the client to the server.
+		// - `is-loading` when the image is being downloaded
+		//    from the server to the client.
+		// - `is-blob` when the image is shown using local `blob` data.
+
+		const classes = classNames(
+			'editor-media-modal-detail__preview',
+			'is-image',
+			{ 'is-uploading': uploading },
+			{ 'is-loading': loading },
+			{ 'is-blog': isBlob },
+		);
+
+		// A fake image element is added behind the preloading image
+		// in order to improve the UX between the states that an image could have,
+		// for instance when the image is restored.
+		const fakeClasses = classNames(
+			'editor-media-modal-detail__preview',
+			'is-image',
+			'is-fake',
+			{ 'is-uploading': uploading },
+			{ 'is-loading': loading },
+			{ 'is-blog': isBlob },
```

Comment:
it isn't used in any place for now but we have to decide if we wanna show the image without a special aspect when the data is ready to use locally (blob), for instance when the image is edited.
Keep in mind that, in this case, while that image is shown immediately, at the same time, it is uploaded (`is-uploading`) and thus an opacity:0.5 is added.